### PR TITLE
Generics

### PR DESCRIPTION
**Pull request by timyhac**
Support for arrays, UDTs, Strings (with different logic depending on CPU)

### CODE REVIEW

**Comment by kyle-github** [comment]
You are a machine!

File: src/libplctag/DataTypes/StringMarshaller.cs [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+{
+    public class StringMarshaller : IMarshaller<string>
+    {
+
+        const int MAX_LOGIX_STRING_LENGTH = 82;
+        const int MAX_PCCC_STRING_LENGTH = 80;
+
+        public PlcType PlcType { get; set; }
+
+        public int? ElementSize
```
Comment (timyhac) [inline src/libplctag/DataTypes/StringMarshaller.cs:14]: @kyle-github - this is one that I'm not sure about.

This works for my CompactLogix, but I don't know how to make this work for other Plc types.

**Comment by timyhac** [comment]
Hmmm, this doesn't work for BOOL arrays.

Might have to rejig encode/decode functions so they encode/decode the entire buffer or object rather than one at a time.

**Comment by timyhac** [comment]
Ok now it supports BOOL arrays as well.

**Pull request merged**